# **Byte Pair Encoding (BPE) - Gelişmiş**


## **Gelişmiş BPE Uygulaması ve Geniş Veri Seti**

Önceki çalışmamızda BPE algoritmasının temel çalışma mantığını küçük bir örnek üzerinde incelemiştik. Bu bölümde ise öğrendiğimiz tüm kavramları birleştirerek algoritmayı çok daha kapsamlı bir veri seti üzerinde uygulayacağız.

Veri setini genişletmekteki temel amacımız, dilin yapısındaki kalıpları daha doğru tespit etmek ve istatistiksel olarak en verimli birleştirme (merge) işlemlerini gerçekleştirmektir. Veri hacmi ne kadar büyürse, algoritmanın sık kullanılan heceleri ve kelimeleri öğrenme başarısı o kadar artar.

Ayrıca bu bölümde, Tokenizer yapısının en önemli iki fonksiyonunu inşa edeceğiz:
1. **Encoding:** Metni, öğrendiğimiz kurallara göre sayısal token dizisine dönüştürme işlemi.
2. **Decoding:** Sayısal tokenları tekrar okunabilir metne dönüştürme işlemi.

### **Veri Seti: Oktay Sinanoğlu ve Çok Elektronlu Teori**

Veri setimizi zenginleştirmek için, **"Türk Einstein'ı"** olarak anılan ve 26 yaşında Yale Üniversitesi'nde "tam profesör" unvanını alarak son 300 yılın en genç profesörü sıfatıyla tarihe geçen **Prof. Dr. Oktay Sinanoğlu**'nun çalışmalarını kullanacağız.

Sinanoğlu, özellikle atom ve moleküllerin elektronik yapılarını ve kimyasal davranışlarını matematiksel olarak açıklayan "Çok Elektronlu Atom Teorisi" ile kuantum kimyasında devrim yaratmıştır. Aşağıdaki veri seti, onun bu alandaki temel taşı kabul edilen akademik makalesinin Türkçe çevirisinden derlenmiştir.

> 🔗 **Kaynak:** Sinanoğlu, O. (1961). *Many-Electron Theory of Atoms and Molecules*. Proceedings of the National Academy of Sciences (PNAS), 47(8), 1217–1226. [Makaleye Erişim (PDF)](https://www.pnas.org/doi/pdf/10.1073/pnas.47.8.1217)

<p align="center">
  <img src="https://github.com/omertarikyilmaz/byte-pair-encoding/blob/main/assets/oktay_sinanoglu.jpg?raw=true" alt="Oktay Sinanoğlu" width="600">
</p>

In [ ]:
# @title Veri Seti "Many-Electron Theory of Atoms and Molecules" (Çift Tıklayarak Açabilirsin)
dataset = "ATOM VE MOLEKÜLLERİN ÇOK ELEKTRONLU TEORİSİ OKTAY SİNANOĞLU YALE ÜNİVERSİTESİ Atom ve moleküllerin orbital teorileri nitel veya yarı deneysel bir şekilde oldukça iyi çalışır ancak nicel bir teste tabi tutulduğunda başarısız olur Hata o kadar büyüktür ki genellikle molekül oluşumu bile tahmin edilemez ve bu durum genellikle elektron korelasyonuna atfedilir Elektronlar birbirlerini sadece Hartree Fock yöntemindeki gibi ortalama potansiyelleriyle değil anlık potansiyelleriyle de etkiledikleri için elimizde bir çok elektron problemi vardır Ancak bu ne tür bir çok cisim problemidir Uzun menzilli Coulomb itmeleri tüm elektronların tek bir karmaşık harekete dahil olmasına mı neden olur Eğer dışlama ilkesinin etkilerini yenecek ölçüde durum bu olsaydı kabuk yapısı yok olurdu ve bir atom veya molekül daha çok bir elektron sıvısı damlasına benzerdi Gerçek kabuk yapısı ve elektron yoğunlukları Hartree Fock yöntemiyle verilenlere yakındır Örneğin ikincisinden elde edilen yoğunluklar X ışını sonuçlarıyla oldukça iyi uyuşur Atomlar ve moleküllerdeki çok elektronlu problemin durumu çekirdek içi ve sonsuz elektron gazı problemleriyle karşılaştırılır Çekirdeklerin içinde güçlü ve kısa menzilli kuvvetler yalnızca yerel nükleon çifti korelasyonlarının baskın olmasına neden olur Brueckner teorisi bunları idealize edilmiş sonsuz nükleer madde için dikkate alır İtmelerin gücü nedeniyle her parçacık sürekli olarak kutuplaştırdığı bir ortamda hareket eder Sonlu sistemlerde zorluklar ortaya çıkar Bu kutuplaşmış deniz potansiyeli parçacık durumuna güçlü bir şekilde bağlıdır bu nedenle ortogonal temel durum orbitalleri kolayca elde edilemez Temel süreksizdir ve sıradan pertürbasyon teorisinde olduğu gibi yavaş yakınsayan sonsuz toplamları değerlendirme zorlukları ortaya çıkar Sonsuz elektron gazında Hartree Fock orbitalleri düzlem dalgaları olduğundan elektron lokalizasyonu yoktur Korelasyon iki elektron arasındaki anlık Coulomb potansiyeli ile birbirlerine uygulayacakları ortalama yani Hartree Fock potansiyeli arasındaki fark tarafından belirlenir Ancak tam delokalizasyon nedeniyle bu Hartree Fock kısmı tekdüzedir ve çok büyük bir kutu için neredeyse sıfırdır ve dalgalanma potansiyeli uzun menzilli tam potansiyeldir Bu nedenle birçok elektron aynı anda korelasyona girer Neyse ki metalik davranışın yaklaşıldığı boyalar vb hariç atomlar ve moleküllerde sorun elektron gazı sorunundan çok farklıdır Simetri çekirdek kuyuları ve dışlama ilkesi eşleşmiş spinlere sahip elektronları Hartree Fock dalga fonksiyonunda lokalize eder Linnett ve Poe tarafından gösterildiği gibi örneğin Neon atomu için bir determinant fonksiyonu tarafından verilen maksimum olasılık konfigürasyonu tetrahedral olarak düzenlenmiş dört elektron çiftidir Bu tür uzaysal düzenlemeler bağlar vb Hartree Fock orbitallerini determinantı değiştirmeyen eşdeğer orbitallere dönüştürerek daha iyi temsil edilir Atomlar ve moleküller üzerine yapılan tek doğru hesaplamalar hala Helyum üzerinde Hylleraas ve Hidrojen molekülü üzerinde James ve Coolidge in klasik çalışmalarına dayanan iki elektronlu sistemler üzerinedir Bu yöntem N elektron durumuna genişletilmediğinden bu tür sistemler için konfigürasyon etkileşimi kullanılır İkincisi genellikle yavaş yakınsar dejenerasyon veya rezonans için kullanıldığı durumlar hariç çoklu uyarımların sayısı N ile çok hızlı artar ve küçük moleküller üzerindeki en iyi konfigürasyon etkileşimi şimdiye kadar bağ enerjilerinin sadece yaklaşık üçte biri ile yarısını vermiştir Helyum ve Hidrojen molekülü durumunun zorluk derecesinin nicel bir şemasını elde etmek ve çok elektronlu hareketin fiziksel davranışını anlamak ve kimyasal resmi korelasyonla ilişkilendirmek için N elektronlu atom veya molekül için kapsamlı bir teori geliştirdik Başlangıç çalışmamızda pertürbasyon teorisinde kesin çözümler elde edildi Şimdi bu sonuçların çok ötesine geçtik ve resmi çeşitli yönlerde tamamladık Bu iletişimde ortaya çıkan genel yapı ve fiziksel temele vurgu yaparak en son bulgularımızı özetliyoruz Tam matematiksel teori Journal of Chemical Physics dergisinde yayınlanacaktır Atom ve moleküllerdeki çok elektronlu problemin temel özellikleri şunlardır Coulomb itmelerinin uzun menzilli etkileri Hartree Fock potansiyeli ve orbitalleri tarafından iyi bir şekilde ele alınır Biz Hartree Fock çözümü ile başlarız Elektronların içinde hareket ettiği bu denizin seçilme nedeni ilerledikçe daha belirgin hale gelecektir Brueckner yönteminde olduğu gibi kutuplaşmış bir deniz kullanmanın etkileri yalnızca oldukça yüksek derecelere kadar ortaya çıkar ve gerektireceği hantal kendi kendine tutarlılık prosedürleri dışında ihmal edilebilir düzeyde kalır Kesin dalga fonksiyonu için tek Hartree Fock determinantı ve korelasyon düzeltmesini yazarsak enerji bir Hartree Fock ve bir korelasyon kısmına ayrılır Dalgalanma potansiyeli yani korelasyon davranışı ikili dalgalanma potansiyelleri ve denizin antisimetrik karakterine güçlü bir şekilde bağlı olan elektron dağılımı tarafından belirlenir Dışlama ilkesi üçüncü ve dördüncü elektronları farklı bir kabuğa koyarak onları diğer elektronların dalgalanma potansiyellerinden güvenli bir mesafede tutar Diğer durumların tartışılmasının da göstereceği gibi bu olgu oldukça geneldir ve daha incelikli durumlarda bile temel etkiler olarak yerel çift korelasyonlarına yol açar ve üç dört elektronlu çarpışmaları genellikle ihmal edilebilir kılar Bu gerçek denklemde kendini gösterir Tam korelasyon fonksiyonu üç dört vb elektronlu kümeleri de içerecektir Ancak açıklanan fiziksel etkiler nedeniyle kümeler aynı uzay bölgesinde büyük olmayacaktır bu nedenle enerjide yalnızca çift enerjileri önemlidir Berilyum atomunun yeni konfigürasyon etkileşimi çalışmasında tekli ve üçlü uyarımlar tamamen ihmal edilebilir bulunmuştur Ancak önemli dörtlü uyarımlar ortaya çıkmıştır ilk bakışta şaşırtıcı bir sonuç çünkü dört elektronlu bir çarpışma olasılığının üç cisimli olandan daha küçük olması gerekir Aslında iki elektron çarpıştığında sistemin başka bir yerinde diğer ikili çarpışmaların bağımsız ancak eşzamanlı olarak gerçekleşmesi çok muhtemeldir Bunları hesaba katmak için bağımsız çift fonksiyonlarının tüm olası ürünleri eklenmelidir Kusurlu gazlar teorisinde olduğu gibi bunlar bağlantısız kümelerdir ve diğer çok cisim problemlerinde önemli bir rol oynamıştır Berilyumda gerçek dört elektronlu korelasyonlar ihmal edilebilirse dörtlü uyarımların katsayıları yalnızca ikili uyarımların katsayıları tarafından verilmelidir Tabloda bu tür katsayılar konfigürasyon etkileşimi ile karşılaştırılmıştır Uyum gerçekten tatmin edicidir Lityum Hidrür üzerindeki son sonuçları inceleyerek benzer sonuçlar elde ettik Bağlantısız kümelerin enerji etkisi diyagramlarla analiz edilir Atomlar ve moleküllerde Berilyumda olduğu gibi rezonans nedeniyle birimden yeterince farklıysa bağlantısız kümeler enerjiyi etkiler Bu enerji değişimi tablodaki sonuçla mükemmel bir şekilde uyuşur ve beklendiği gibi ihmal edilebilir olduğunu gösterir Çok parçacıklı etkilerin küçüklüğünün bir nedeni korelasyon fonksiyonunun küçüklüğüdür Genellikle denklem iyi bir yaklaşım olacaktır ancak gerekirse ihmal edilen terimler de tahmin edilebilir ve enerji için bir üst sınır elde edilebilir Berilyum ve Lityum Hidrürde olduğu gibi elektronlar radyal olarak farklı uzay bölgelerinde lokalize edildiğinde üç dört ve daha fazla elektron korelasyonunun neden önemsiz hale geldiğini gördük Peki ya Bor atomundaki gibi hepsi radyal olarak aynı civarda olan elektronları içeren durumlar ne olacak Bu tür durumlarda üç elektronlu korelasyonların nasıl davranmasının beklendiğini görmek için Bor atomunu inceliyoruz Berilyumdaki korelasyonun neredeyse tamamı karışımından kaynaklanmaktadır Ancak Bor atomunda olduğu gibi kabuğun üzerine bir elektron konulduğunda korelasyon ona ortogonal hale gelir Tüm küre yerine korelasyon artık yalnızca düzlem civarında büyük olacaktır Üç elektronlu korelasyonu yani üçgeni kayda değer kılmak için çarpım büyük olmalıdır Ancak dalgalanma potansiyelinin büyük olduğu yerde büyüktür Dördüncü elektron beşinci elektronun en olası yerinde olduğunda gördüğü dalgalanma potansiyeli sabittir Çarpım her yerde küçük olacak ve yine çok elektronlu korelasyonları önemsiz kılmak için dalgalanma potansiyeli ve dışlamanın birleşik davranışını sergileyecektir Bir çift enerjileri toplamına sahip olduğumuzda her çift fonksiyonu ayrı ayrı elde edilebilir Bu yollar ve teorimizden çıkan genel resim şöyledir Her bir çift yörünge ortogonalliğine tabi olarak değiştirildiğinde Hartree Fock denizindeki elektron çiftlerinin Schrödinger denklemlerini verir Bu denklem Bethe Goldstone Brueckner denkleminin kapalı formundan potansiyel yerine dalgalanma potansiyeli içermesiyle ayrılır Bu bağlamda elde ettiğimiz sonuçlar sonlu çekirdekler için de yararlı olmalıdır Şimdi varyasyonel minimum prensibi her bir çifte ayrı ayrı uygulanabilir ve Helyum ve Hidrojen molekülünde olduğu gibi güçlü bir korelasyon deliği olan kabuklar için açıkça terimler içeren deneme fonksiyonları kullanılabilir Berilyumda olduğu gibi rezonanslı olanlar gibi diğerleri için birkaç konfigürasyonu karıştıran küçük bir seküler denklemin çözülmesi gerekir İhtiyaç duyulan bağımsız çift fonksiyonlarının sayısı birçok çift durumunun çokluğu nedeniyle N değerinden çok daha küçüktür Neon için kırk beş yerine altısı tekli olmak üzere sadece dokuz çift fonksiyonu elde edilmelidir Yörünge ortogonalliği nedeniyle minimizasyon prosedürü otuzların kuantum kimyasını rahatsız eden iç kabukların kabusu olan yanlış kabuğa yakınsama korkusu olmadan dış kabuk çiftlerine bile uygulanabilir Dış bir çift için denklem iki elektronlu Hamiltonyen formuna sahiptir ve bu tür yarı deneysel şemalar için bir temel sağlar Dahası M O hesaplamalarının başarılarını açıklayan denkleme eklenebilir Güçlü elektron delokalizasyonu olmayan moleküllerdeki Hartree Fock orbitalleri eşdeğer orbitallere dönüştürülebilir O zaman bağ korelasyon enerjilerinin ve bağlanmamış bölgeler arasındaki Van der Waals çekimlerinin toplamı olur İkincisi Pitzer tarafından yarı deneysel olarak tartışılmıştır Teori her iki şekilde de kullanılabilir ve Hartree Focktan başlayarak korelasyon sonuçlarına sahip bu tür Van der Waals kuvvetlerini hesaplayabilir veya bağ enerjileri ve London dağılım tipi kuvvetlerin tahminlerinden H F orbitalleri çiftlerinin korelasyon enerjilerini tahmin edebilir Böylece korelasyona rağmen bağlar Lewis Langmuir oktetleri ve orbitalleri ile kimyasal resim varlığını sürdürür"

## **Temel Fonksiyonlar**

Aşağıdaki kod bloğunda tanımladığımız `get_stats` ve `merge` fonksiyonları, **`bpe-temeller.ipynb`** dosyasında detaylıca incelediğimiz ve algoritmanın çekirdeğini oluşturan fonksiyonların **birebir aynısıdır.**

Mantık değişmemiştir, sadece bu fonksiyonları şimdi daha büyük bir veri seti üzerinde kullanacağız:

* **`get_stats(ids)`:** Veri setini tarayarak yan yana gelen byte ikililerinin frekansını hesaplar (**İstatistikçimiz**).
* **`merge(ids, pair, idx)`:** Belirlenen en sık çifti listede bulur ve yeni atanan token ID'si ile değiştirerek sıkıştırmayı yapar (**Birleştiricimiz**).

In [ ]:
def get_stats(ids):
    counts = {}
    for pair in zip(ids, ids[1:]):
        counts[pair] = counts.get(pair, 0) + 1
    return counts

def merge(ids, pair, idx):
  newids = []
  i = 0
  while i < len(ids):
    if i < len(ids) - 1 and ids[i] == pair[0] and ids[i+1] == pair[1]:
      newids.append(idx)
      i += 2
    else:
      newids.append(ids[i])
      i += 1
  return newids


##**Hiperparametre Tanımları ve Veri Hazırlığı**

Bu bölüm, BPE algoritmasının eğitim sınırlarını çizdiğimiz ve veriyi işleme hazırladığımız konfigürasyon aşamasıdır.

1.  **`vocab_size = 276` (Hedef Kelime Dağarcığı):**
    * Standart UTF-8 kodlamasında **0-255** arası değerler temel byte'ları (karakter parçalarını) temsil eder.
    * Belirlediğimiz **276** değeri, algoritmanın bu temel setin üzerine **20 adet yeni "birleştirilmiş token"** ekleyerek duracağını belirtir. Bu değer artırıldıkça metin daha fazla sıkışır, ancak kelime dağarcığı (vocabulary) büyür.

2.  **`num_merges` (İterasyon Sayısı):**
    * Eğitim döngüsünün kaç adım çalışacağını belirleyen sayaçtır.
    * Formül: $\text{Hedef Boyut} - \text{Temel Boyut} = 276 - 256 = 20 iterasyon.


In [ ]:
vocab_size = 276
num_merges = vocab_size - 256
ids = list(dataset.encode("utf-8"))

merges = {}
for i in range(num_merges):
  stats = get_stats(ids)
  pair = max(stats, key=stats.get)
  idx = 256 + i

  print(f"{pair} çifti birleştirilerek yeni token {idx} oluşturuluyor")

  ids = merge(ids, pair, idx)
  merges[pair] = idx

print(f"\nText Uzunluğu: {len(dataset)}, Token Uzunluğu: {len(list(dataset.encode('utf-8')))}")
print(f"Sıkıştırma Oranı: {len(list(dataset.encode('utf-8'))) / len(ids):.2f}X")

(196, 177) çifti birleştirilerek yeni token 256 oluşturuluyor
(108, 101) çifti birleştirilerek yeni token 257 oluşturuluyor
(114, 32) çifti birleştirilerek yeni token 258 oluşturuluyor
(110, 32) çifti birleştirilerek yeni token 259 oluşturuluyor
(108, 97) çifti birleştirilerek yeni token 260 oluşturuluyor
(101, 32) çifti birleştirilerek yeni token 261 oluşturuluyor
(195, 188) çifti birleştirilerek yeni token 262 oluşturuluyor
(105, 32) çifti birleştirilerek yeni token 263 oluşturuluyor
(195, 167) çifti birleştirilerek yeni token 264 oluşturuluyor
(107, 32) çifti birleştirilerek yeni token 265 oluşturuluyor
(197, 159) çifti birleştirilerek yeni token 266 oluşturuluyor
(97, 32) çifti birleştirilerek yeni token 267 oluşturuluyor
(111, 110) çifti birleştirilerek yeni token 268 oluşturuluyor
(105, 258) çifti birleştirilerek yeni token 269 oluşturuluyor
(196, 159) çifti birleştirilerek yeni token 270 oluşturuluyor
(100, 101) çifti birleştirilerek yeni token 271 oluşturuluyor
(97, 114) çifti 

##**Decoding ve Encoding (Çözme ve Kodlama)**

Eğitim aşamasını tamamladık ve `merges` sözlüğünde kurallarımızı oluşturduk. Şimdi bu kuralları kullanarak Tokenizer'ın asıl işlevlerini inşa edeceğiz.

1.  **Decoding (Kod Çözme):** Sayısal token listesini alıp tekrar okunabilir metne dönüştürme işlemidir.
    * Eğitimden gelen `merges` sözlüğü sadece birleştirmeleri (rules) tutar. Decoding için her ID'nin (örn: 256) karşılık geldiği byte dizisini (örn: `b't' + b'h'`) tutan kapsamlı bir **`vocab` (Kelime Dağarcığı)** tablosuna ihtiyacımız vardır.

2.  **Encoding (Kodlama):** Ham metni alıp, öğrendiğimiz birleştirme kurallarını sırasıyla uygulayarak sıkıştırılmış token listesine dönüştürme işlemidir.
    * Metin önce ham byte'lara (0-255) çevrilir.
    * Ardından `merges` kurallarına göre taranır; eğer kurallarda geçen bir ikili (pair) bulunursa, o ikili yeni ID ile değiştirilir. Bu işlem yapılabilecek birleştirme kalmayana kadar sürer.

<p align="center">
  <img src="https://github.com/omertarikyilmaz/byte-pair-encoding/blob/main/assets/tokenizer_logic.png?raw=true" alt="Tokenizer Mantığı" width="800">
</p>

##**Decoding (Kod Çözme)**

Decoding, eğitim aşamasında elde ettiğimiz sayısal verileri (tokenları) tekrar insan tarafından okunabilir metne dönüştürme sürecidir. Bu işlem, kodlamanın (encoding) aksine karmaşık aramalar gerektirmez; deterministiktir ve basit bir "eşleştirme" mantığına dayanır.



**Süreç üç temel adımda gerçekleşir:**

1.  **Kelime Dağarcığının (Vocabulary) İnşası:**
    * Önce 0-255 arasındaki temel byte değerleri hafızaya alınır.
    * Ardından, eğitim sırasında öğrendiğimiz "birleştirme kuralları" sırasıyla uygulanır. Örneğin; eğer eğitimde `(100, 200)` çifti `256` numaralı token olduysa, sistem `256`'nın değerinin bu iki byte'ın birleşimi olduğunu kaydeder. Böylece her soyut ID'nin (256, 257...) gerçek dünyadaki byte karşılığı (örn: "mer", "haba") belirlenmiş olur.

2.  **Eşleştirme ve Birleştirme:**
    * Elimizdeki sayısal token listesindeki her ID'nin karşılığı bu sözlükten çekilir.
    * Elde edilen küçük byte parçaları uç uca eklenerek tek ve uzun bir ham veri bloğu (stream) oluşturulur.

3.  **Metne Dönüştürme (Render):**
    * Oluşan ham byte dizisi, UTF-8 standardına göre yorumlanarak tekrar anlamlı metne (string) çevrilir. Eğer bozuk bir byte varsa, metin akışını bozmamak adına yerine özel bir "bilinmeyen" karakteri () konulur.

In [ ]:
vocab = {idx: bytes([idx]) for idx in range(256)}
for (p0, p1), idx in merges.items():
    vocab[idx] = vocab[p0] + vocab[p1]

def decode(ids):
  tokens = b"".join(vocab[idx] for idx in ids)
  text = tokens.decode("utf-8", errors="replace")
  return text

In [ ]:
print(decode([274]))

in


###**Encoding (Kodlama)**

Encoding, ham metni (string) alıp, eğitim sırasında öğrendiğimiz `merges` (birleştirme kuralları) sözlüğünü kullanarak sıkıştırılmış sayısal token listesine dönüştürme işlemidir.

**Kodun İşleyiş Mantığı:**

1.  **Ham Veriye Dönüş:** Metin önce UTF-8 standardına göre ham byte listesine (`[109, 101, 114...]`) çevrilir.
2.  **Döngüsel Tarama (`while`):** Liste üzerinde birleştirilecek hiçbir çift kalmayana kadar (veya liste boyutu 1'e düşene kadar) döngü çalışır.
3.  **En İyi Adayı Seçme (`min` mantığı):**
    * Mevcut listedeki tüm ikililerin istatistiği (`stats`) çıkarılır.
    * Bu ikililer arasından `merges` sözlüğünde var olan ve **en küçük Token ID'sine sahip olan** çift seçilir.
    * *Neden `min`?* Çünkü kuralları, eğitim sırasındaki öncelik sırasına göre (256, sonra 257...) uygulamamız gerekir. İlk öğrendiğimiz kural, en temel kuraldır.
4.  **Birleştirme:** Seçilen çift listede bulunur ve karşılık gelen yeni Token ID ile değiştirilir.


Bu işlem sonucunda uzun bir byte dizisi, `merges` kurallarına göre sıkıştırılmış daha kısa bir token listesine dönüşür.

In [ ]:
def encode(text):
  tokens = list(text.encode("utf-8"))
  while len(tokens) >= 2:
    stats = get_stats(tokens)
    pair = min(stats, key=lambda p: merges.get(p, float("inf")))
    if pair not in merges:
      break

    idx = merges[pair]
    tokens = merge(tokens, pair, idx)
  return tokens

In [ ]:
print(encode("Tam matematiksel teori Journal of Chemical Physics dergisinde yayınlanacaktır."))

[84, 97, 109, 32, 109, 97, 116, 101, 109, 97, 116, 105, 107, 115, 101, 108, 32, 116, 101, 111, 114, 263, 74, 111, 117, 114, 110, 97, 108, 32, 111, 102, 32, 67, 104, 101, 109, 105, 99, 97, 108, 32, 80, 104, 121, 115, 105, 99, 115, 32, 271, 114, 103, 105, 115, 274, 100, 261, 121, 97, 121, 256, 110, 260, 110, 97, 99, 97, 107, 116, 256, 114, 46]


In [ ]:
print(decode(encode("Tam matematiksel teori Journal of Chemical Physics dergisinde yayınlanacaktır")))

Tam matematiksel teori Journal of Chemical Physics dergisinde yayınlanacaktır


###**Doğrulama ve Test (Validation)**

Tokenizer'ımızın güvenilirliğini kanıtlamak için bir "Gidiş-Dönüş" (Round-Trip) testi yapıyoruz.

**Testin Mantığı:**
Bir metni önce sayılara çevirip (`encode`), ardından o sayılardan tekrar metne dönüştürdüğümüzde (`decode`), elimize geçen sonucun **orijinal metinle birebir aynı** olması gerekir.

* **Girdi:** Orijinal metin (`valtext`).
* **Süreç:** Metin -> Tokenlar (ID) -> Metin
* **Sonuç:** `Orijinal Metin == İşlem Görmüş Metin` kıyaslaması.

Eğer çıktı **`True`** ise, algoritmanız veriyi **kayıpsız (lossless)** bir şekilde işliyor ve hiçbir bilgiyi (boşluk, özel karakter vb.) yolda kaybetmiyor demektir.

In [ ]:
valtext = "Prof. Dr. Oktay Sinanoğlu, 26 yaşında Yale Üniversitesi'nde profesör unvanını alarak modern zamanların en genç profesörü olmuştur. Kuantum kimyası, moleküler biyoloji ve matematik alanlarındaki çalışmalarıyla dünya çapında tanınan Sinanoğlu, aynı zamanda Türkçe'nin bilim dili olarak korunması için büyük mücadele vermiştir."

valtext2 = decode(encode(valtext))

print(f"Orijinal Metin: {valtext}")
print(f"İşlem Görmüş Metin: {valtext2}")
print("=======================================================")
print(f"Kayıpsız Dönüşüm Başarılı mı? : {valtext2 == valtext}")

Orijinal Metin: Prof. Dr. Oktay Sinanoğlu, 26 yaşında Yale Üniversitesi'nde profesör unvanını alarak modern zamanların en genç profesörü olmuştur. Kuantum kimyası, moleküler biyoloji ve matematik alanlarındaki çalışmalarıyla dünya çapında tanınan Sinanoğlu, aynı zamanda Türkçe'nin bilim dili olarak korunması için büyük mücadele vermiştir.
İşlem Görmüş Metin: Prof. Dr. Oktay Sinanoğlu, 26 yaşında Yale Üniversitesi'nde profesör unvanını alarak modern zamanların en genç profesörü olmuştur. Kuantum kimyası, moleküler biyoloji ve matematik alanlarındaki çalışmalarıyla dünya çapında tanınan Sinanoğlu, aynı zamanda Türkçe'nin bilim dili olarak korunması için büyük mücadele vermiştir.
Kayıpsız Dönüşüm Başarılı mı? : True
